tests/test_serialization.py

```diff
@@ -0,0 +1,189 @@
+import pytest
+import copy
+import pickle
+import tenseal as ts
+
+from tests.utils import *
+
+
+def ctx():
+    return ts.context(ts.SCHEME_TYPE.CKKS, 8192, 0, [60, 40, 40, 60])
+
+
+def deep_copy(ctx):
+    return copy.deepcopy(ctx)
+
+
+def simple_copy(ctx):
+    return copy.copy(ctx)
+
+
+def internal_copy(ctx):
+    return ctx.copy()
+
+
+def recreate(ctx):
+    proto = ctx.serialize()
+    return ts.context_from(proto)
+
+
+def pickled(ctx):
+    out = pickle.dumps(ctx)
+    return pickle.loads(out)
+
+
+@pytest.mark.parametrize("duplicate", [deep_copy, simple_copy, internal_copy, recreate, pickled,])
+def test_context_recreation(duplicate):
+    orig_context = ctx()
+    context = duplicate(orig_context)
+
+    assert context.is_private() is True, "TenSEALContext should be private"
+    assert context.public_key() is not None, "TenSEALContext shouldn't be None"
+
+    orig_context.make_context_public(generate_galois_keys=False, generate_relin_keys=False)
+    context = duplicate(orig_context)
+    assert context.is_public() is True, "TenSEALContext should be public"
+
```

Comment:
```suggestion

    orig_context = ctx()
    context = duplicate(orig_context)
    context.make_context_public(generate_galois_keys=False, generate_relin_keys=False)
    assert orig_context.is_private() is True, "original TenSEALContext should stay private"
```

Reply:
additional check here would be great, in order to verify that the context object is copied properly.